src/chips/nrf51822/gpio.rs

```diff
@@ -0,0 +1,121 @@
+use core::mem;
+use core::ops::{Index, IndexMut};
+use hil;
+
+// Source: https://github.com/hackndev/zinc/tree/master/volatile_cell
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct VolatileCell<T> {
+    value: T,
+}
+
+#[allow(dead_code)]
+impl<T> VolatileCell<T> {
+    #[inline]
+    pub fn get(&self) -> T {
+        unsafe {
+            ::core::intrinsics::volatile_load(&self.value)
+        }
+    }
+
+    #[inline]
+    pub fn set(&self, value: T) {
+        unsafe {
+            ::core::intrinsics::volatile_store(&self.value as *const T as *mut T, value)
+        }
+    }
+}
+
+#[allow(non_snake_case)]
+struct GPIO {
+    _pad0: [u8; 1284],
+    pub OUT: VolatileCell<u32>,
+    pub OUTSET: VolatileCell<u32>,
+    pub OUTCLR: VolatileCell<u32>,
+    pub IN: VolatileCell<u32>,
+    pub DIR: VolatileCell<u32>,
+    pub DIRSET: VolatileCell<u32>,
+    pub DIRCLR: VolatileCell<u32>,
+    _pad1: [u8; 480],
+    pub PIN_CNF: [VolatileCell<u32>; 32],
+}
+
+#[allow(non_snake_case)]
+fn GPIO() -> &'static GPIO {
```

Comment:
The connonical way of doing this is:

```rust
impl GPIO {
    fn new() -> &'static GPIO {
        unsafe { mem::transmute(0x50000000 as usize) }
    }
}
```

However, this might be a nice alternative. As I'm suggesting, the obvious thing to do would be to stick the result in a static variable, at which point the contents has to remain an unsafe pointer because we can't call `mem::transmute` from statics (yet anyway).

This actually ends up being a reasonably nice interface.